### PR TITLE
Tests: Adds reporter that shows JS test errors on GitHub when executing GitHub workflows

### PR DESCRIPTION
**Pull request by ockham**
## Description
Try adding a Jest Reporter that produces an output format for GitHub annotations. 

Based on https://github.com/facebook/jest/pull/11320 (but ported the TypeScript to plain ol' JS since I couldn't get the TS based reporter to work). Quoting that PR:

> Add a Jest reporter to output a format that will be interpreted by GitHub Actions as annotations -- i.e. any errors reported by Jest will show up in a PR's diff below the offending line. This should make for significantly improved Developer Experience, since developers no longer have to scan CI logs for errors, but instead will be notified of them at the location where they happen.
 
This PR is mostly meant to demo that feature; ideally, we'll be able to tweak it and merge it. If https://github.com/facebook/jest/pull/11320 gets merged, we can then consider switching to that, once there's a Jest version that includes it.

## How has this been tested?
See diff view: There should be an annotation below the failing assertion.

To test locally, run
```
GITHUB_ACTIONS=true npm run test-unit -- test/unit/test/index.test.js
```

to see the GHA-specific output format. Confirm that if you omit the `GITHUB_ACTIONS` env var, you won't see that output.

## Screenshots

image

---

FYI @gziolo

### CODE REVIEW

**Comment by gziolo** [comment]
There is a bug in the reporter which means we need to wait more to see it in action:

> /home/runner/work/gutenberg/gutenberg/test/unit/GithubActionsReporter.js:14
> 		const messages = getMessages( _aggregatedResults?.testResults );
> 		                                                 ^
> SyntaxError: Unexpected token '.'

I'm waiting patiently 😄

**Comment by ockham** [comment]
> There is a bug in the reporter which means we need to wait more to see it in action:
> 
> > /home/runner/work/gutenberg/gutenberg/test/unit/GithubActionsReporter.js:14
> > const messages = getMessages( _aggregatedResults?.testResults );
> > ^
> > SyntaxError: Unexpected token '.'
> 
> I'm waiting patiently 😄

Yeah, I think it's the `?.` which isn't supported by Node 12. The [Node 14 one](https://github.com/WordPress/gutenberg/pull/31041/checks?check_run_id=2401226915) isn't bothered by that (and the unit test fails as it should), but the error reporter doesn't make it to the GH diff 😕

**Comment by ockham** [comment]
Seems like we _are_ actually generating annotations: https://github.com/WordPress/gutenberg/actions/runs/771200933 They just aren't displayed in the diff 🤔 

Wonder if we're doing something wrong with regard to how we pass the filename and/or line and column...

**Comment by ockham** [comment]
Ah, I see it. We're extracting the wrong line number. From my local run:

```
Ran all test suites matching /test\/unit\/test\/index.test.js/i.
::error file=/Users/ockham/src/gutenberg/test/unit/test/index.test.js,line=93,col=5::Error: expect(received).toBe(expected) // Object.is equality%0A%0AExpected: "b"%0AReceived: "a"%0A    at Object.<anonymous> (/Users/ockham/src/gutenberg/test/unit/test/index.test.js:3:17)%0A    at Object.asyncJestTest (/Users/ockham/src/gutenberg/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:106:37)%0A    at /Users/ockham/src/gutenberg/node_modules/jest-jasmine2/build/queueRunner.js:45:12%0A    at new Promise (<anonymous>)%0A    at mapper (/Users/ockham/src/gutenberg/node_modules/jest-jasmine2/build/queueRunner.js:28:19)%0A    at /Users/ockham/src/gutenberg/node_modules/jest-jasmine2/build/queueRunner.js:75:41%0A    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

Note that we're setting `line=93,col=5` -- which is coming from the _last_ line in the stack trace:

```
at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

whereas the correct location would be `line=3,col=17`, based on

```
at Object.<anonymous> (/Users/ockham/src/gutenberg/test/unit/test/index.test.js:3:17)
```

I'll try to fix the RegEx.

**Comment by ockham** [comment]
💥 

image

TODO:
- [x] Let's see if we can revert 52a24a2
- [ ] ~Can we use ANSI colors? See https://github.com/actions/toolkit/blob/e76decaf8a52bc981d17ddd3efcca548f6bc1a0e/packages/core/README.md#styling-output~ _Punted, see https://github.com/WordPress/gutenberg/pull/31041#issuecomment-824740362_
- [x] Disable for local runs, see https://github.com/facebook/jest/pull/11320#issuecomment-823808159; _but_: this might actually come in handy for editors such as VS Code. 68dfa7e

**Review by gziolo — APPROVED** [review]
Awesome, feel free to experiment a bit more but don't let folks wait too long to see it in action 🎉

**Comment by ockham** [comment]
On the note of ANSI colors: I've noticed that Jest provides those as part of the items in the `failureMessages` array, e.g.

```
[
  {
    ancestorTitles: [ 'My test' ],
    duration: 5,
    failureDetails: [ [Object] ],
    failureMessages: [
      'Error: \x1B[2mexpect(\x1B[22m\x1B[31mreceived\x1B[39m\x1B[2m).\x1B[22mtoBe\x1B[2m(\x1B[22m\x1B[32mexpected\x1B[39m\x1B[2m) // Object.is equality\x1B[22m\n' +
        '\n' +
        'Expected: \x1B[32m"b"\x1B[39m\n' +
        'Received: \x1B[31m"a"\x1B[39m\n' +
        '    at Object.<anonymous> (/Users/bernhardreiter/src/gutenberg/test/unit/test/index.test.js:3:17)\n' +
        '    at Object.asyncJestTest (/Users/bernhardreiter/src/gutenberg/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:106:37)\n' +
        '    at /Users/bernhardreiter/src/gutenberg/node_modules/jest-jasmine2/build/queueRunner.js:45:12\n' +
        '    at new Promise (<anonymous>)\n' +
        '    at mapper (/Users/bernhardreiter/src/gutenberg/node_modules/jest-jasmine2/build/queueRunner.js:28:19)\n' +
        '    at /Users/bernhardreiter/src/gutenberg/node_modules/jest-jasmine2/build/queueRunner.js:75:41\n' +
        '    at processTicksAndRejections (internal/process/task_queues.js:93:5)'
    ],
    fullName: 'My test A equals B',
    location: null,
    numPassingAsserts: 0,
    status: 'failed',
    title: 'A equals B'
  }
]
```

However, the seem to get lost during processing in the reporter 🤔

**Comment by ockham** [comment]
I'll ignore the colors for now. I think that it's not specific to this reporter -- Jest normally uses colors to highlight test outcomes in its standard reporters, and those are apparently also lost (stripped by Jest?) in the GHA output.

**Comment by gziolo** [comment]
Will it work also with e2e tests? it would be more helpful there as they fail more often 😅

**Comment by ockham** [comment]
> Will it work also with e2e tests? it would be more helpful there as they fail more often 😅

Ah, great question! I think it should, since they're also Jest-based. However, we'll need to add the reporter to the e2e tests' `jest.config.js`, and probably move it to a different location.

I'll give it a quick spin 👍

**Comment by ockham** [comment]
Pushed a few more commits, and moved the reporter to the `scripts` package, so it could be part of the "base" Jest configs for e2e (and unit) tests. Also pushed a (temporary) commit with an e2e test failure to demo that here.

@gziolo Care to give it another quick look if the changes I made to files in `packages/scripts/` look okay?

File: packages/e2e-tests/specs/editor/various/inserting-blocks.test.js [hunk 1]
```diff
@@ -112,7 +112,7 @@ describe( 'Inserting blocks', () => {
 
 		// Verify vertical traversal at offset. This has been buggy in the past
 		// where verticality on a blank newline would skip into previous block.
-		await page.keyboard.type( 'Foo' );
+		await page.keyboard.type( 'Bar' );
```
Comment (gziolo) [inline packages/e2e-tests/specs/editor/various/inserting-blocks.test.js]:
```suggestion
		await page.keyboard.type( 'Foo' );
```

**Comment by gziolo** [comment]
> Care to give it another quick look if the changes I made to files in packages/scripts/ look okay?

Sure thing, everything looks good. I'm very excited about this integration!

I made a screenshot and reverted the enforced test failure:

Screen Shot 2021-04-23 at 08 20 10

**Comment by ockham** [comment]
> I made a screenshot and reverted the enforced test failure

Awesome, thank you very much!

**Pull request merged**